File: apps/gate/gate/modelcall.py
"""/model_call - the only path from any agent to a model (DESIGN.md sec. 4).

Capability-checked, budget-capped per run, and every call is cost-logged into
agent_calls whether it succeeds or fails.
"""

from typing import Any

from .capabilities import CAPABILITIES
from .db import pool
from .models_ai import azure_chat, resolve_tier
from .pipeline import GateError


def handle_model_call(req: dict[str, Any]) -> dict[str, Any]:
    agent_name = req.get("agentName")
    run_id = req.get("runId")
    messages = req.get("messages")

    allowed = CAPABILITIES.get(agent_name or "")
    if not allowed or "model_call" not in allowed:
        raise GateError(403, f"agent '{agent_name}' has no capability 'model_call'")
    if not run_id:
        raise GateError(400, "model calls must belong to a run")
    if not isinstance(messages, list) or not messages:
        raise GateError(400, "messages required")

    try:
        tier = resolve_tier(agent_name, req.get("tier"))
    except ValueError as err:
        raise GateError(403, str(err)) from err

    # Budget check against the administrator-approved cap for this run.
    with pool.connection() as conn:
        budget = conn.execute(
            """SELECT r.approved_budget_microdollars AS budget,
                      COALESCE((SELECT SUM(cost_microdollars) FROM agent_calls WHERE run_id = r.id), 0)
                    + COALESCE((SELECT SUM(provider_cost_microdollars) FROM actions WHERE run_id = r.id), 0)
                        AS spent
                 FROM runs r WHERE r.id = %s""",
            (run_id,),
        ).fetchone()
    if budget is None:
        raise GateError(404, f"run {run_id} not found")
    cap, spent = int(budget["budget"]), int(budget["spent"])
    if spent >= cap:
        raise GateError(
            402, f"run {run_id} exhausted its budget (spent {spent} of {cap} microdollars)"
        )

    with pool.connection() as conn:
        agent_call = conn.execute(
            """INSERT INTO agent_calls (run_id, task_id, agent_name, model_id, model_tier, status)
               VALUES (%s, %s, %s, 'pending', %s, 'started') RETURNING id""",
            (run_id, req.get("taskId"), agent_name, tier),
        ).fetchone()
    agent_call_id = agent_call["id"]

    try:
        result = azure_chat(
            tier,
            messages,
            json_mode=bool(req.get("json")),
            max_tokens=req.get("maxTokens"),
        )
    except Exception as err:
        with pool.connection() as conn:
            conn.execute(
                "UPDATE agent_calls SET status = 'failed', completed_at = now() WHERE id = %s",
                (agent_call_id,),
            )
        raise GateError(502, f"model call failed: {err}") from err

    with pool.connection() as conn:
        conn.execute(
            """UPDATE agent_calls
                  SET model_id = %s, input_tokens = %s, cached_input_tokens = %s,
                      output_tokens = %s, cost_microdollars = %s, status = 'completed',
                      completed_at = now()
                WHERE id = %s""",
            (
                result["model_id"],
                result["input_tokens"],
                result["cached_input_tokens"],
                result["output_tokens"],
                result["cost_microdollars"],
                agent_call_id,
            ),
        )

    return {
        "agentCallId": str(agent_call_id),
        "content": result["content"],
        "modelId": result["model_id"],
        "tier": tier,
        "inputTokens": result["input_tokens"],
        "cachedInputTokens": result["cached_input_tokens"],
        "outputTokens": result["output_tokens"],
        "costMicrodollars": result["cost_microdollars"],
        "runSpentMicrodollars": str(spent + result["cost_microdollars"]),
    }

File: apps/gate/src/modelcall.ts
// /model_call - the only path from any agent to a model (DESIGN.md sec. 4).
// Capability-checked, budget-capped per run, and every call is cost-logged
// into agent_calls whether it succeeds or fails.
import { pool } from "./db.js";
import { GateError } from "./pipeline.js";
import { CAPABILITIES } from "./capabilities.js";
import { azureChat, resolveTier, type ChatMessage } from "./models.js";

export interface ModelCallRequest {
  runId: string;
  taskId?: string | null;
  agentName: string;
  tier?: string;
  messages: ChatMessage[];
  json?: boolean;
  maxTokens?: number;
}

export interface ModelCallResult {
  agentCallId: string;
  content: string;
  modelId: string;
  tier: string;
  inputTokens: number;
  cachedInputTokens: number;
  outputTokens: number;
  costMicrodollars: number;
  runSpentMicrodollars: string;
}

export async function handleModelCall(req: ModelCallRequest): Promise<ModelCallResult> {
  const allowed = CAPABILITIES[req.agentName];
  if (!allowed || !allowed.includes("model_call")) {
    throw new GateError(403, `agent '${req.agentName}' has no capability 'model_call'`);
  }
  if (!req.runId) throw new GateError(400, "model calls must belong to a run");
  if (!Array.isArray(req.messages) || req.messages.length === 0) {
    throw new GateError(400, "messages required");
  }

  let tier;
  try {
    tier = resolveTier(req.agentName, req.tier);
  } catch (err) {
    throw new GateError(403, err instanceof Error ? err.message : String(err));
  }

  // Budget check against the administrator-approved cap for this run.
  const budget = await pool.query<{ budget: string; spent: string }>(
    `SELECT r.approved_budget_microdollars AS budget,
            COALESCE((SELECT SUM(cost_microdollars) FROM agent_calls WHERE run_id = r.id), 0)
          + COALESCE((SELECT SUM(provider_cost_microdollars) FROM actions WHERE run_id = r.id), 0)
            AS spent
       FROM runs r WHERE r.id = $1`,
    [req.runId],
  );
  if (!budget.rows[0]) throw new GateError(404, `run ${req.runId} not found`);
  const { budget: cap, spent } = budget.rows[0];
  if (BigInt(spent) >= BigInt(cap)) {
    throw new GateError(
      402,
      `run ${req.runId} exhausted its budget (spent ${spent} of ${cap} microdollars)`,
    );
  }

  const inserted = await pool.query<{ id: string }>(
    `INSERT INTO agent_calls (run_id, task_id, agent_name, model_id, model_tier, status)
     VALUES ($1, $2, $3, 'pending', $4, 'started') RETURNING id`,
    [req.runId, req.taskId ?? null, req.agentName, tier],
  );
  const agentCallId = inserted.rows[0]!.id;

  try {
    const result = await azureChat(tier, req.messages, {
      json: req.json,
      maxTokens: req.maxTokens,
    });
    await pool.query(
      `UPDATE agent_calls
          SET model_id = $2, input_tokens = $3, cached_input_tokens = $4,
              output_tokens = $5, cost_microdollars = $6, status = 'completed',
              completed_at = now()
        WHERE id = $1`,
      [
        agentCallId,
        result.modelId,
        result.inputTokens,
        result.cachedInputTokens,
        result.outputTokens,
        result.costMicrodollars,
      ],
    );
    const newSpent = (BigInt(spent) + BigInt(result.costMicrodollars)).toString();
    return {
      agentCallId,
      content: result.content,
      modelId: result.modelId,
      tier,
      inputTokens: result.inputTokens,
      cachedInputTokens: result.cachedInputTokens,
      outputTokens: result.outputTokens,
      costMicrodollars: result.costMicrodollars,
      runSpentMicrodollars: newSpent,
    };
  } catch (err) {
    await pool.query(
      `UPDATE agent_calls SET status = 'failed', completed_at = now() WHERE id = $1`,
      [agentCallId],
    );
    throw new GateError(502, `model call failed: ${err instanceof Error ? err.message : String(err)}`);
  }
}
